### PR TITLE
fix method lookup intersection error

### LINKED ISSUE #23024: Method lookup with `Tuple{DataType,Any}` regression
Reduced from OpenCL.jl test failure.

```julia
function f2(::Type{T}, ::Int) where T
    1 + 1
end
@show code_typed(f2, Tuple{DataType,Any})

v = 1
T = UInt8
g() = f2(T::DataType, v)

@code_warntype g()
g()
```

Output:
```
code_typed(f2, Tuple{DataType, Any}) = 0-element Array{Any,1}
Variables:
  #self#::#g

Body:
  begin
      return (Main.f2)((Core.typeassert)(Main.T, Main.DataType)::DataType, Main.v)::Union{}
  end::Union{}
Unreachable reached at 0x7fb886a9814e
```

Doesn't happen on 0.6.

### PR DESCRIPTION
**Pull request by vtjnash**
fix #23024 and improve subtype test macro to always check for this

### CODE REVIEW

File: test/subtype.jl [hunk 1]
```diff
@@ -600,15 +600,25 @@ function test_properties()
 end
 
 macro testintersect(a, b, result)
-    if isa(result,Expr) && result.head === :call && length(result.args)==2 && result.args[1] === :!
+    if isa(result, Expr) && result.head === :call && length(result.args) == 2 && result.args[1] === :!
         result = result.args[2]
-        cmp = :notequal_type
```
Comment (JeffBezanson) [inline test/subtype.jl]: These functions can be useful for instrumenting by editing the definitions at the top of the file.

File: src/subtype.c [hunk 2]
```diff
@@ -2142,20 +2142,9 @@ jl_value_t *jl_type_intersection_env_s(jl_value_t *a, jl_value_t *b, jl_svec_t *
...
             if (szb > 0 && !jl_types_equal(b, (jl_value_t*)jl_type_type)) {
-                if (jl_subtype_env(*ans, b, env, szb)) {
-                    if (jl_is_leaf_type(*ans)) {
-                        for(i=0; i < sz; i++) {
-                            if (jl_is_typevar(env[i])) {
-                                *ans = jl_bottom_type; goto bot;
-                            }
-                        }
-                    }
-                }
```
Comment (JeffBezanson) [inline src/subtype.c]: We should watch out for inference performance regressions here due to looking at more methods.

File: test/subtype.jl [hunk 3]
```diff
@@ -600,15 +600,25 @@ function test_properties()
...
+    b = esc(b)
+    result = esc(result)
     Base.remove_linenums!(quote
-        @test $(esc(cmp))(_type_intersect($(esc(a)), $(esc(b))), $(esc(result)))
-        @test $(esc(cmp))(_type_intersect($(esc(b)), $(esc(a))), $(esc(result)))
+        # test real intersect
+        @test $cmp(_type_intersect($a, $b), $result)
+        @test $cmp(_type_intersect($b, $a), $result)
+        # test simplified intersect
+        if !($result === Union{})
```
Comment (martinholters) [inline test/subtype.jl]: `&& $(cmp === :(==))`?

File: test/subtype.jl (hunk 1 above)
Comment (vtjnash) [inline test/subtype.jl]: hm, but there's also special handling for using a function named `==` in `@test` which allows it to print more detailed information on failure

File: test/subtype.jl (hunk 3 above)
Comment (vtjnash) [inline test/subtype.jl]: Currently those end up being the same. I think I'll let the next person worry about making this more precise, if they ever need that case.

File: src/subtype.c (hunk 2 above)
Comment (vtjnash) [inline src/subtype.c]: @nanosoldier `runbenchmarks(ALL, vs=":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/23058#discussion_r130705704) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/13ca47c664bb68fbf77fa82f85ffe9241c501e0e/fef5313_vs_fb76529/report.md). cc @ararslan

File: src/subtype.c (hunk 2 above)
Comment (vtjnash) [inline src/subtype.c]: It looks like CI and nanosoldier times are stable at least.

**Comment by JeffBezanson** [comment]
If this works out I guess we should backport it eventually.

**Pull request merged**